### PR TITLE
WT-9913 Add example for tiered storage.

### PR DESCRIPTION
**Pull request by ddanderson**

### CODE REVIEW

File: examples/c/ex_tiered.c [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+    error_check(session->open_cursor(session, uri, NULL, NULL, &cursor));
+
+    while ((ret = cursor->next(cursor)) == 0) {
+        cursor->get_key(cursor, &key);
+        cursor->get_value(cursor, &value);
+        printf(" %d: %s\n", key, value);
+    }
+
+    /* We expect "not found" to end the iteration.  Anything else is an error. */
```
Comment (sueloverso) [inline examples/c/ex_tiered.c:75]: This may or may not be true as WT_ROLLBACK can be returned if we revert the changes in WT-10133. I agree with the code though.

File: examples/c/ex_tiered.c [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+#define BUILD_DIR "../../"
+#define STORAGE_SOURCE "dir_store"
+
+int
+main(int argc, char *argv[])
+{
+    WT_CONNECTION *conn;
+    WT_SESSION *session;
+    const char *home;
+    char config[1024], bucketdir[1024];
```
Comment (sueloverso) [inline examples/c/ex_tiered.c:99]: Alphabetize

File: examples/c/ex_tiered.c [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+    /* Configure the connection to use tiered storage. */
+    error_check(wiredtiger_open(home, NULL, config, &conn));
+
+    /* Open a session for the current thread's work. */
+    error_check(conn->open_session(conn, NULL, NULL, &session));
+
+    /* Create a table that lives locally. Tiered storage is disabled for this file. */
+    error_check(session->create(
+      session, "table:housecat", "key_format=i,value_format=S,tiered_storage=(name=none)"));
```
Comment (sueloverso) [inline examples/c/ex_tiered.c:137]: First I would macro-ize the `table:` URIs throughout the program. Maybe URI_LOCAL and URI_TIERED or something. Second, `s_all` does not like `housecat` and flags it for spelling. While I understand you were trying to not be boring with names, probably something like `table:local_data` and `table:tiered_data` or similar keeps things very clear.

File: examples/c/ex_tiered.c [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+    /* Create a table using tiered storage. */
+    error_check(session->create(session, "table:wildcat", "key_format=i,value_format=S"));
+
+    /* Add entries to both tables. */
+    add_data(session, "table:housecat", 0, 10);
+    add_data(session, "table:wildcat", 0, 10);
```
Comment (sueloverso) [inline examples/c/ex_tiered.c:144]: You use 10 in all the places, macro-ize it to NUM_ENTRIES. I could see someone wanting to change this to using lots of entries as a simple test for larger objects or something like that.

File: examples/c/ex_tiered.c [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+
+    /*
+     * Do a regular checkpoint. Checkpoints are usually done in their own thread with their own
+     * session. Data is synchronized to local storage.
+     */
+    error_check(session->checkpoint(session, NULL));
+
+    /* Add more entries to both tables. */
+    add_data(session, "table:housecat", 10, 10);
```
Comment (sueloverso) [inline examples/c/ex_tiered.c:153]: Similarly, the first key should be `NUM_ENTRIES` and the next one `2 * NUM_ENTRIES` so that if someone wants to add a million keys each time, it all just works.

File: examples/c/ex_tiered.c [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+
+    /*
+     * Do a tiered checkpoint. For tiered tables, new data is flushed (synchronized) to the
+     * configured tiered storage.
+     */
+    error_check(session->checkpoint(session, "flush_tier=(enabled)"));
+
+    /* Show the data. */
+    show_data(session, "table:housecat", "local table after adding 20 items");
+    show_data(session, "table:wildcat", "tiered table after adding 20 items and flush_tier call");
```
Comment (sueloverso) [inline examples/c/ex_tiered.c:164]: Similarly I suggest getting rid of the specific number in the message part and sprintf'ing it into a buffer with the macro to have it be correct.

File: examples/c/ex_tiered.c [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+     */
+    error_check(session->checkpoint(session, NULL));
+
+    /*
+     * In the wildcat table, some of the entries (up to key 19), has been put into tiered storage,
+     * and the rest is backed by a local file. However, all queries on the data look the same.
+     */
+    show_data(session, "table:housecat", "local table after adding 10 more items");
+    show_data(session, "table:wildcat",
+      "tiered table after adding 10 more items that have not been flushed");
```
Comment (sueloverso) [inline examples/c/ex_tiered.c:181]: Same here about not putting hard values in the string.

File: examples/c/ex_tiered.c [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+    add_data(session, "table:housecat", 20, 10);
+    add_data(session, "table:wildcat", 20, 10);
+
+    /*
+     * Another regular checkpoint. No new data is flushed to tiered storage.
+     */
+    error_check(session->checkpoint(session, NULL));
+
+    /*
+     * In the wildcat table, some of the entries (up to key 19), has been put into tiered storage,
```
Comment (sueloverso) [inline examples/c/ex_tiered.c:176]: I suggest changing 'wildcat' to 'tiered' and adjusting the comment to avoid saying '19'.

**Review by sueloverso — APPROVED** [review]
I have a bunch of minor comments. But don't need to re-review unless you'd like me to do so.

**Review by pmacko86 — APPROVED** [review]
LGTM. I really appreciate the detailed comments!

File: examples/c/ex_tiered.c (hunk 1 above)
Comment (ddanderson) [reply examples/c/ex_tiered.c:82]: I don't think we should change this "not found" comment until or unless WT-10133 is reverted.  If we do revert WT-10133 (I'm opposed to this!) we'll need to change a lot of comments and a lot of code in lots of examples.

**Comment by ddanderson** [comment]
@raviprakashgiri29, can you please take a quick look at this example code?

**Review by raviprakashgiri29 — APPROVED** [review]
